call.go

```diff
@@ -63,6 +63,18 @@ func Invoke(ctx context.Context, method string, args, reply interface{}, cc *Cli
 var unaryStreamDesc = &StreamDesc{ServerStreams: false, ClientStreams: false}
 
 func invoke(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, opts ...CallOption) error {
+	for i, opt := range opts {
+		if unaryIntOpt, ok := opt.(UnaryClientInterceptorCallOption); ok {
+			// Filter out the current interceptor from the list of call options. This ensures
+			// that a per-call interceptor sees only its successors, not itself or its predecessors
+			// in the list of call options passed to it.
+			fopts := make([]CallOption, 0, len(opts)-1)
+			fopts = append(append(fopts, opts[:i]...), opts[i+1:]...)
+
+			return unaryIntOpt.UnaryClientInterceptor(ctx, method, req, reply, cc, invoke, fopts...)
```

Comment:
This is not exactly what I had in mind, but I think it could be okay and that what I did have in mind would be too hard to implement.

What I was thinking is that we'd process the call options one-by-one, applying them to the RPC's settings.  When we encountered an interceptor, we'd call it with the remaining, unprocessed `CallOption`s, and it would call an invoker that remembered the state applied from the processed `CallOptions`.

Thoughts on this from a behavior / implementation perspective?  @erni27 / @easwars?

Reply:
Hm, correct me if I'm wrong but it means that a particular interceptor wouldn't see `CallOption`s defined before itself and `CallOption`s are processed before that interceptor is even invoked, doesn't it?
Indeed it would be harder to implement but the other thing is that it seems a bit too complex to use. I mean, as a user I expect that an interceptor (in general so it doesn't matter if this is a `DialOption` interceptor or a `CallOption` one) intercepts the call itself so is invoked before any `CallOption` is processed. It would make the users think to not only about the ordering of interceptors itself but also about the ordering of all applied `CallOption`s and how they're processed by subsequent interceptors.
I don't have a strong opinion on that. Your proposal seems to be correct from `CallOption`s point of view.

Reply:
@erni27 

We had a lengthy discussion about this and we've reached an impasse. There seem to be multiple options here, but none of them are straightforward and none of them seem to cover all bases. We discussed coming up with a different type of call option interceptor, which would *not* accept call options in its arguments. We also discussed the case where we want one call option interceptor to be able to add another call option interceptor. But given that the place where we chain and call interceptors is different to the place where we process call options, implementing this introduces quite a bit of complexity.

Until we have some users clearly asking for a certain type of call option interceptor, we've decided to put this on hold. It was a mistake on our part to have you start working on this. But I guess it took someone to work on this for us to realize that this was not all that well scoped out.

We really apologize for taking up your time and energy on this.

We hope you will continue to contribute to your repo through other issues/PRs. I'm closing this PR for now, but feel free to let us know if you have other thoughts on this one. Thank you.